lib/public-stats.js

```diff
@@ -101,37 +102,108 @@ const updateIndexerQueryStats = async (pgClient, committees) => {
 /**
  * @param {pg.Client} pgClient
  * @param {Iterable<Committee>} committees
+ * @param {(minerId: string, cid: string) => (string[] | undefined)} findDealClients
  */
-const updateDailyDealsStats = async (pgClient, committees) => {
-  let total = 0
-  let indexed = 0
-  let retrievable = 0
+const updateDailyDealsStats = async (pgClient, committees, findDealClients) => {
+  /** @type {Map<string, Map<string, {
+   * tested: number;
+   * majority_found: number;
+   * indexed:  number;
+   * indexed_http: number;
+   * retrievable: number;
+   * }>>} */
+  const minerClientDealStats = new Map()
   for (const c of committees) {
-    total++
+    const { minerId, cid } = c.retrievalTask
+    let clients = findDealClients(minerId, cid)
+    if (!(clients && clients.length)) {
+      console.warn(`Invalid retrieval task (${minerId}, ${cid}): no deal clients found. Using a placeholder value 'no-clients'.`)
```

Comment:
what if we instead kept the `clients` array empty?

Reply:
Then, we will not insert any records into the table because we need both the miner & client for that. As a result, we will exclude deals where we don't have any client info from the DRS calculation.

Having written that, it should never happen that there is a task with no deal clients. 

The table `retrievable_deals` which tracks deals eligible for retrievability testing uses `(cid, miner_id)` as the primary key:

https://github.com/filecoin-station/spark-api/blob/2712b4e04c37f6759fb832701b3b6b2afb9fa001/migrations/052.do.retrievable-deals-clients.sql#L5

Per-round retrieval tasks are defined as a random sample from that table, so each task must have at least one deal client:

https://github.com/filecoin-station/spark-api/blob/2712b4e04c37f6759fb832701b3b6b2afb9fa001/api/lib/round-tracker.js#L260-L273

I want to handle this unlikely edge case to avoid losing measurements. Even if we cannot link them to any deal client, we can still link them to a miner (an SP).

Reply:
> Having written that, it should never happen that there is a task with no deal clients.

Shall we log this situation to Sentry then?

Reply:
Added Sentry logging in 6424760

Reply:
What kind of inconsistency do you have in mind?

If we abort the operation, we discard measurements that provide valid information about a storage provider.

Having written that, we should never get an empty list of clients, so I am fine with skipping such a committee. We will get a Sentry report in the unlikely case that this happens, I think that shoud be good enough 👍🏻

Reply:
See c3f43f6

Reply:
If `'no-clients'` makes it into the DB, I would consider that an inconsistency: You expect a client to be there if a row / column value is there.